File: python/servicer.py
import asyncio
import websockets

class Server:
    def __init__(self, header_size=64):
        self.welcome_message = b'@w|h:' + str(header_size).encode()
        self.header_size = header_size
        self.functions = {}
        self.bytes_read = 0

    async def __get_job_info(request):
        packet_name = None
        packet_size = None
        packet_id = None

        sections = request[0:self.header_size + 1].decode('ascii').split('|')

        if sections[0] == '@j':
            parameters = sections[1].split(',')

            for parameter in parameters:
                parameter_tuple = parameter.split(':')

                if parameter_tuple[0] == 'n':
                    packet_name = parameter_tuple[1]
                elif parameter_tuple[0] == 's':
                    packet_size = int(parameter_tuple[1])
                elif parameter_tuple[0] == 'd':
                    packet_id = parameter_tuple[1]
                else:
                    print('Invalid header parameter sent')

        return (packet_name, packet_size, packet_id)

    async def __get_inputs(request, packet_name):
        input_schema = self.functions[packet_name]['inputs']
        offset = 0
        bytes_read = 0
        inputs = {}

        for i, input in enumerate(input_schema):
            start_pos = self.header_size + 1 + offset + (i * self.header_size)
            end_pos = start_pos + self.header_size + 1
            parameters = request[start_pos:end_pos].decode('ascii').split(',')
            bytes_read += header_size

            input_name = None
            input_type = None
            input_size = None

            for parameter in parameters:
                parameter_tuple = parameter.split(':')

                if parameter_tuple[0] == 'i':
                    input_name = parameter_tuple[1]
                elif parameter_tuple[0] == 't':
                    input_type = parameter_tuple[1]
                elif parameter_tuple[0] == 's':
                    input_size = int(parameter_tuple[1])
                else:
                    print('Invalid input parameter sent')
                    return (None, None)

            if input_name and input_type and input_size:
                if input_name not in input_schema or input_type != input_schema[input_name]:
                    print('Invalid input')
                    return (None, None)

                offset += input_size + 1
                bytes_read += input_size

                raw_input_data = request[end_pos:input_size + 1]
                input_data = None

                if input_type == 'string':
                    input_data = raw_input_data.decode('ascii')
                if input_type == 'binary':
                    input_data = raw_input_data
                if input_type == 'integer':
                    input_data = int(raw_input_data)
                if input_type == 'float':
                    input_data = float(raw_input_data)
                if input_type == 'boolean':
                    input_data = bool(raw_input_data)

                if input_data:
                    inputs[input_name] = input_data
                else:
                    print('Invalid input type')
                    return(None, None)
            else:
                print('Missing input parameter')
                return (None, None)

        return (inputs, bytes_read)

    async def __build_output_packet(self, outputs, packet_name, packet_id):
        output_schema = self.functions[packet_name]['outputs']
        packet_header = b'@r|n:' + packet_name.encode() + b',d:' + packet_id.encode()
        output_packet = b''

        for output_name, raw_value in outputs.items():
            if output_name in output_schema:
                type_schema = output_name[output_schema]
                output_value = None
                if type_schema == 'string':
                    if type(raw_value) is str:
                        output_value = raw_value.encode()
                    else:
                        print('Output value does not match schema')
                        return None
                if type_schema == 'binary':
                    if type(raw_value) is bytes:
                        output_value = raw_value
                    else:
                        print('Output value does not match schema')
                        return None
                if type_schema == 'integer':
                    if type(raw_value) is int:
                        output_value = str(raw_value).encode()
                    else:
                        print('Output value does not match schema')
                        return 
                if type_schema == 'float':
                    if type(raw_value) is float:
                        output_value = str(raw_value).encode()
                    else:
                        print('Output value does not match schema')
                        return None
                if type_schema == 'boolean':
                    if type(raw_value) is bool:
                        output_value = str(int(raw_value)).encode()
                    else:
                        print('Output value does not match schema')
                        return None
                else:
                    print('Output schema invalid')
                    return None

                output_header = b'o:' + output_name.encode()
                output_header += b',t:' + output_schema[output_name].encode()
                output_header += b',s:' + str(len(output_value)).encode()

                if len(output_header) > self.header_size:
                    print('Output header too big')
                    return None
                else:
                    output_packet += output_header.ljust(self.header_size) + output_value
            else:
                print('Invalid function output')
                return None

        packet_header += b',s:' + str(len(output_packet) + self.header_size).encode()

        if len(packet_header) > self.header_size:
            print('Packer header too big')
            return None
        else:
            output_packet = packet_header.ljust(self.header_size) + output_packet
            return output_packet

    async def __process_incoming_message(self, request):
        packet_name, packet_size, packet_id = await self.__get_job_info(request)
        if not packet_name or not packet_size or not packet_id:
            return b'@e|m:invalid_job_request'

        inputs, bytes_read = await self.__get_inputs(request, packet_name)
        if not inputs or bytes_read:
            return b'@e|m:invalid_inputs'

        if (bytes_read + self.header_size != packet_size):
            return b'@e|m:invalid_packet_size'

        outputs = self.functions[packet_name]['function'](**inputs)
        output_packet = await self.__build_output_packet(outputs, packet_name, packet_id)
        if not output_packet:
            return b'@e|m:server_function_error'

        return output_packet

    def __add_to_welcome_message(self, function_def):
        self.welcome_message += b'|n:' + function_def['name'].encode()
        for input, input_type in function_def['inputs'].items():
            if input_type not in ['string', 'boolean', 'integer', 'float', 'binary']:
                print('Invalid input type')
            self.welcome_message += b';i:' + input.encode() + b',t:' + input_type.encode()
        for output, output_type in function_def['outputs'].items():
            self.welcome_message += b';o:' + output.encode() + b',t:' + output_type.encode()

    async def __handler(self, websocket, _):
        await websocket.send(self.welcome_message)
        while True:
            request = await websocket.recv()
            response = await self.__process_incoming_message(request)
            await websocket.send(response)

    def add_function(self, **kwargs):
        name = kwargs.get('name', None)
        if not name:
            print('Function name must be provided')

        inputs = kwargs.get('inputs', None)
        if not inputs:
            print('Function inputs must be provided')

        outputs = kwargs.get('outputs', None)
        if not outputs:
            print('Function outputs must be provided')

        function = kwargs.get('function', None)
        if not function:
            print('Function method must be provided')

        function_def = {
            'name': name,
            'inputs': inputs,
            'outputs': outputs,
            'function': function}

        self.functions[name] = function_def
        self.__add_to_welcome_message(function_def)

    def start_server(self, ip, port):
        asyncio.get_event_loop().run_until_complete(websockets.serve(self.__handler, ip, port, max_size=None))
        asyncio.get_event_loop().run_forever()

File: node/servicer.js
const WebSocket = require('ws');
const shortid = require('shortid');

class Client {
  constructor(serverList) {
    this.callbacks = {};
    this.functions = {};
    this.isConnected = false;
    this.serverList = serverList;
    shuffleArray(this.serverList);
  }

  _readResponseHeader(response) {
    const responseHeader = binaryToString(response.subarray(0, this.headerSize)).split('|');
    if (responseHeader[0] === '@r') {
      const responseHeaderParams = responseHeader[1].split(',');
      let responseId = null;
      let responseName = null;
      let responseSize = null;

      for (let i = 0; i < responseHeaderParams.length; i++) {
        const param = responseHeaderParams[i].split(':');

        if (param[0] === 'n') {
          responseName = param[1].trim();
        } else if (param[0] === 's') {
          responseSize = parseInt(param[1].trim(), 10);
        } else if (param[0] === 'd') {
          responseId = param[1].trim();
        }
      }

      if (responseId && responseName && responseSize) {
        return {
          id: responseId,
          name: responseName,
          size: responseSize
        };
      } else {
        return {
          error: 'Server sent invalid response'
        };
      }
    } else {
      return {
        error: 'Server sent invalid response'
      };
    }
  }

  _readResponseParams(response, responseHeader) {
    const outputSchema = this.functions[responseHeader.name].outputs;
    const outputs = {};
    let offset = 0;
    let bytesRead = 0;

    for (let i = 0; i < Object.keys(outputSchema).length; i++) {
      const startPos = this.headerSize + offset + (i * this.headerSize);
      const endPos = startPos + this.headerSize;
      const parameters = binaryToString(response.subarray(startPos, endPos)).split(',');
      bytesRead += this.headerSize;

      let outputName = null;
      let outputType = null;
      let outputSize = null;

      for (let j = 0; j < parameters.length; j++) {
        const param = parameters[j].split(':');

        if (param[0] === 'o') {
          outputName = param[1].trim();
        } else if (param[0] === 's') {
          outputSize = parseInt(param[1].trim(), 10);
        } else if (param[0] === 't') {
          outputType = param[1].trim();
        }
      }

      if (outputName && outputType && outputSize) {
        if (!outputSchema.hasOwnProperty(outputName) || outputType !== outputSchema[outputName]) {
          return {
            error: 'Invalid input parameter'
          };
        }

        offset += outputSize;
        bytesRead += outputSize;

        const rawOutputData = response.subarray(endPos, endPos + outputSize);
        let outputData = null;

        if (outputType === 'string') {
          outputData = binaryToString(rawOutputData).trim();
        } else if (outputType === 'integer') {
          outputData = parseInt(binaryToString(rawOutputData).trim(), 10);
        } else if (outputType === 'float') {
          outputData = parseFloat(binaryToString(rawOutputData).trim());
        } else if (outputType === 'boolean') {
          outputData = (binaryToString(rawOutputData).trim() === '1');
        } else if (outputType === 'binary') {
          outputData = rawOutputData;
        }

        if (outputData) {
          outputs[outputName] = outputData;
        } else {
          return {
            error: 'Invalid response type'
          };
        }
      } else {
        return {
          error: 'Missing response parameter'
        };
      }
    }

    return {
      outputs: outputs,
      bytesRead: bytesRead
    };
  }

  _responseHandler(response, flags) {
    if (flags.binary) {
      const responseHeader = this._readResponseHeader(response);
      if (responseHeader && responseHeader.hasOwnProperty('error')) {
        console.log(responseHeader['error']);
        return;
      } else if (!responseHeader) {
        console.log('Invalid response format');
        return;
      }

      let callback = null;
      if (this.callbacks.hasOwnProperty(responseHeader.id)) {
        callback = this.callbacks[responseHeader.id];
      } else {
        console.log('Invalid response ID');
        return;
      }

      let error = null;
      const responseParams = this._readResponseParams(response, responseHeader);
      if (responseParams && responseParams.hasOwnProperty(error)) {
        error = responseParams.error;
      } else if (!responseParams) {
        error = 'Invalid response';
      }

      if (responseParams.bytesRead !== responseHeader.size) {
        error = 'Invalid packet size';
      }

      let callbackParams = null;
      if (!error && responseParams.hasOwnProperty('outputs')) {
        callbackParams = responseParams.outputs;
      }
      callback(error, callbackParams);
      delete this.callbacks[responseHeader.id];
    } else {
      console.log('Server sent invalid response');
    }
  }

  _decodeHeader(header) {
    const decodedData = binaryToString(header).split('|');

    if (decodedData[0] === '@w') {
      const headerSizeParam = decodedData[1].split(':');

      if (headerSizeParam[0] === 'h') {
        this.headerSize = parseInt(headerSizeParam[1], 10);

        for (let i = 2; i < decodedData.length; i++) {
          const functionMultiParams = decodedData[i].split(';');
          const functionInputs = {};
          const functionOutputs = {};
          let functionName = '';

          for (let j = 0; j < functionMultiParams.length; j++) {
            const functionParams = functionMultiParams[j].split(',');
            const multiParam = {};

            for (let k = 0; k < functionParams.length; k++) {
              const functionParam = functionParams[k].split(':');
              multiParam[functionParam[0]] = functionParam[1];
            }

            if (multiParam.hasOwnProperty('i') && multiParam.hasOwnProperty('t')) {
              const inputName = multiParam['i'];
              const inputType = multiParam['t'];
              functionInputs[inputName] = inputType;
            } else if (multiParam.hasOwnProperty('o') && multiParam.hasOwnProperty('t')) {
              const outputName = multiParam['o'];
              const outputType = multiParam['t'];
              functionOutputs[outputName] = outputType;
            } else if (multiParam.hasOwnProperty('n')) {
              functionName = multiParam['n'];
            } else {
              console.log('Invalid header parameter received')
            }
          }

          if (functionName && functionInputs && functionOutputs) {
            this.functions[functionName] = {
              inputs: functionInputs,
              outputs: functionOutputs
            };
          } else {
            console.log('Invalid function definition received');
          }
        }
      } else {
        console.log('Invalid header size parameter received');
      }
    } else {
      console.log('Invalid welcome message received');
    }
  }

  connect(cb) {
    for (let i = 0; i < this.serverList.length; i++) {
      this.socketErrorCount = 0;

      const that = this;
      const ip = this.serverList[i][0];
      const port = this.serverList[i][1];

      let ws = new WebSocket(`ws://${ip}:${port.toString()}`);

      ws.on('error', err => {
        console.log(`[ws://${err.address}:${err.port}] - ${err.code}`);
        this.socketErrorCount++;
        if (this.socketErrorCount === this.serverList.length) {
          cb('No servers reachable', null);
        }
      });

      ws.on('message', function receiveHeader(data, flags) {
        ws.removeListener('message', receiveHeader);

        if (!that.isConnected && flags.binary) {
          that.isConnected = true;
          this.socketErrorCount = 0;
          console.log(`[${ws.url}] - Connected`);
          that._decodeHeader(data);

          ws.on('message', that._responseHandler.bind(that));
          that.ws = ws;

          cb(null, {
            url: ws.url,
            functions: that.functions
          });
        } else {
          this.socketErrorCount++;
        }

        ws = null;
      });
    }
  }

  callFunction(functionName, functionInputs, cb) {
    if (this.functions.hasOwnProperty(functionName)) {
      const packetId = shortid.generate();
      let packetHeader = `@j|n:${functionName},d:${packetId}`;
      let packetBody = null;

      this.callbacks[packetId] = cb;

      const functionInputKeys = Object.keys(functionInputs);
      for (let i = 0; i < functionInputKeys.length; i++) {
        const inputName = functionInputKeys[i];
        const rawInputValue = functionInputs[inputName];

        if (this.functions[functionName].inputs.hasOwnProperty(inputName)) {
          const inputSchema = this.functions[functionName].inputs[inputName];
          let inputValue = null;

          if (inputSchema === 'string') {
            if (typeof rawInputValue === 'string') {
              inputValue = stringToBinary(rawInputValue, rawInputValue.length);
            } else {
              cb('Invalid input type', null);
              return;
            }
          } else if (inputSchema === 'integer') {
            if (typeof rawInputValue === 'number' && Number.isInteger(rawInputValue)) {
              const intStr = rawInputValue.toString();
              inputValue = stringToBinary(intStr, intStr.length);
            } else {
              cb('Invalid input type', null);
              return;
            }
          } else if (inputSchema === 'float') {
            if (typeof rawInputValue === 'number' && !Number.isInteger(rawInputValue)) {
              const floatStr = rawInputValue.toString();
              inputValue = stringToBinary(floatStr, floatStr.length);
            } else {
              cb('Invalid input type', null);
              return;
            }
          } else if (inputSchema === 'boolean') {
            if (typeof rawInputValue === 'boolean') {
              const boolStr = (rawInputValue | 0).toString();
              inputValue = stringToBinary(boolStr, boolStr.length);
            } else {
              cb('Invalid input type', null);
              return;
            }
          } else if (inputSchema === 'binary') {
            if (typeof rawInputValue === 'object') {
              inputValue = rawInputValue;
            } else {
              cb('Invalid input type', null);
              return;
            }
          }

          const inputSize = inputValue.length;
          const inputHeaderStr = `i:${inputName},t:${inputSchema},s:${inputSize}`;
          const inputHeader = stringToBinary(inputHeaderStr, this.headerSize);
          const inputBuffer = concatBuffers(inputHeader, inputValue);

          if (packetBody) {
            packetBody = concatBuffers(packetBody, inputBuffer);
          } else {
            packetBody = inputBuffer;
          }
        } else {
          cb('Invalid input type', null);
          return;
        }
      }

      const packetSize = packetBody.length;
      packetHeader += `,s:${packetSize}`;
      packetHeader = stringToBinary(packetHeader, this.headerSize);

      const packet = concatBuffers(packetHeader, packetBody);
      this.ws.send(packet, {
        binary: true,
        mask: true
      });
    } else {
      cb('Invalid function input', null);
      return;
    }
  }

}

function shuffleArray(a) {
  for (let i = a.length; i; i--) {
    const j = Math.floor(Math.random() * i);
    [a[i - 1], a[j]] = [a[j], a[i - 1]];
  }
}

function concatBuffers(buffer1, buffer2) {
  const newBuffer = new Uint8ClampedArray(buffer1.length + buffer2.length);
  newBuffer.set(buffer1);
  newBuffer.set(buffer2, buffer1.length);
  return newBuffer;
}

function binaryToString(buf) {
  return String.fromCharCode.apply(null, new Uint8ClampedArray(buf));
}

function stringToBinary(str, padding) {
  if (str.length > padding) {
    console.error(`Header exceeds ${padding} bytes`);
  }

  const buf = new Uint8ClampedArray(padding);
  for (let i = 0; i < str.length; i++) {
    const charCode = str.charCodeAt(i);

    if (charCode > 127) {
      console.error('Text must be ASCII');
    } else {
      buf[i] = charCode;
    }
  }

  return buf;
}

module.exports = {
  Client: Client
}
